# Exercício 1

a função $f(x) = x^4 + 4x^3 + 6x^2 + 3x -10$ possui um zero positivo e um zero negativo. Pede-se:

#### a) Encontre, pelo Método da Bissecção, o zero positivo de $f$, partindo de um intervalo de comprimento 1, com precisão de $10^{-3}$

In [79]:
import pandas as pd

x = [-20, -5, -3, -2,-1, 0, 1, 5, 20]
f_x = [j**4 + 4*(j**3) + 6*(j**2) + 3*j - 10 for j in x]
tabela = {'x': pd.Series(data = x), 'f(x)': pd.Series(data = f_x)}
tabela = pd.DataFrame(tabela)
tabela.style.hide_index()

Pela tabela acima, é possível ver que o zero negativo está entre $[-3, -2]$ e o zero positivo está entre $[0,1]$. Como só queremos o zero positivo, usaremos o segundo intervalo. No código abaixo se tem uma função usada para executar o método da bissecção e gerar uma tabela.

In [80]:
import pandas as pd


def bissecao(f, a, b, precisao, max_iteracoes):
    '''
    - Função: bissecao(f, a, b, precisao, max_iteracoes)
    - Autor: Eduardo Freire Mangabeira
    - Obtém um zero de uma função pelo método da bisseção.
    - Feito a partir das aulas do professor Hamilton e do livro Burden, R. L., & Faires, D. J. (1985). Numerical analysis.
    - É precisso passar uma expressão lambda para o parâmetro 'f', fiz isso para deixar o código mais geral e menos verboso.
    - Para expresões lambda, visitar: http://www.dsc.ufcg.edu.br/~pet/jornal/maio2013/materias/tutoriais.html.
    '''
    i = 1
    f_a = f(a)
    p = a + (b-a)/2
    f_p = f(p)
    dados = {'n': [], 'a_n': [], 'b_n': [], 'x_n': [], 'f(a_n)': [], 'f(b_n)': [], 'f(x_n)': []}
    tabela = pd.DataFrame(dados)
    while i <= max_iteracoes:
        p = a + (b-a)/2
        f_p = f(p)
        f_b = f(b)
        dados_2 = {'n': [i], 'a_n': [a], 'b_n': [b], 'x_n': [p], 'f(a_n)': [f_a], 'f(b_n)': [f_b], 'f(x_n)': [f_p]}
        tabela_2 = pd.DataFrame(dados_2)
        tabela = tabela.append(tabela_2, ignore_index = True)
        if f_p == 0 or (b-a)/2 < precisao:
            print("O valor de x que aproxima ao zero da função é: {}".format(p))
            return tabela
        i = i + 1
        if f_a * f_p > 0:
            a = p
            f_a = f_p
        else:
            b = p

In [42]:
print(bissecao.__doc__)


    - Função: bissecao(f, a, b, precisao, max_iteracoes)
    - Autor: Eduardo Freire Mangabeira
    - Obtém um zero de uma função pelo método da bisseção.
    - Feito a partir das aulas do professor Hamilton e do livro Burden, R. L., & Faires, D. J. (1985). Numerical analysis.
    - É precisso passar uma expressão lambda para o parâmetro 'f', fiz isso para deixar o código mais geral e menos verboso.
    - Para expresões lambda, visitar: http://www.dsc.ufcg.edu.br/~pet/jornal/maio2013/materias/tutoriais.html.
    


Agora basta usar os valores dados na questão, considerando o intervalo $[0,1]$.

In [81]:
# f(x) = x^4 + 4(x^3) + 6(x^2) + 3^x - 10
f = lambda x:  x**4 + 4*(x**3) + 6*(x**2) + 3*x - 10
a = 0
b = 1
precisao = 1/(10**3)
max_iteracoes = 20
tabela = bissecao(f, a, b, precisao, max_iteracoes)
tabela.style.hide_index()

O valor de x que aproxima ao zero da função é: 0.8564453125


### Resposta

Pela tabela, vemos que o zero da função com a precisão especificada é encontrado na décima iteração e é dada por $x =  0.8564453125$

#### b)  Repita o exercício anterior partindo de um intervalo de comprimento 0,4

Primeiro vamos checar onde está o zero positivo para um intervalo de 0,4.

In [82]:
x = [0.5, 0.9]
f_x = [j**4 + 4*(j**3) + 6*(j**2) + 3*j - 10 for j in x]
tabela = {'x': pd.Series(data = x), 'f(x)': pd.Series(data = f_x)}
tabela = pd.DataFrame(tabela)
tabela.style.hide_index()

O intervalo $[0.5, 0.9]$ é um bom candidato. Vamos executar o método da bisseção com esses novos parâmetros

In [83]:
f = lambda x:  x**4 + 4*(x**3) + 6*(x**2) + 3*x - 10
a = 0.5
b = 0.9
precisao = 1/(10**3)
max_iteracoes = 20
tabela = bissecao(f, a, b, precisao, max_iteracoes)
tabela.style.hide_index()

O valor de x que aproxima ao zero da função é: 0.85546875


### Resposta

Foi necessária uma iteração a menos nesse caso! O zero da função encontrado:  $0.85546875$.

 
#### c) Encontre, pelo Método de Newton, o zero positivo de f. Use como critério de parada $|x_n - x_{n-1}| < 10^{-4}$.

#### Use 7 casas decimais nos cálculos e arredondamento.

Precisamos escrever uma função também para o método de newton. Também é preciso escrever a derivada de uma função na sua forma  numérica(diferenças finitas).

In [84]:
import pandas as pd


def newton(f, chute_inicial, precisao, max_iteracoes):
    '''
    - Função: newton(f, chute_inicial, precisao, max_iteracoes)
    - Autor: Eduardo Freire Mangabeira
    - Obtém um zero de uma função pelo método de Newton.
    - Feito a partir das aulas do professor Hamilton e do livro Burden, R. L., & Faires, D. J. (1985). Numerical analysis.
    - É precisso passar uma expressão lambda para o parâmetro 'f', fiz isso para deixar o código mais geral e menos verboso.
    - Para expresões lambda, visitar: http://www.dsc.ufcg.edu.br/~pet/jornal/maio2013/materias/tutoriais.html.
    '''
    i = 1
    derivada_numerica = lambda x: (f(x+precisao)-f(x))/ precisao
    dados = {'n':[], 'x_n':[], 'f(x_n)':[], 'f\'x_n':[], '|x_n - x_n-1|':[]}
    tabela = pd.DataFrame(dados)
    while i<= max_iteracoes:
        solucao = chute_inicial - f(chute_inicial)/derivada_numerica(chute_inicial)
        dados2 = {'n':[i], 'x_n':[chute_inicial], 'f(x_n)':[f(chute_inicial)], 'f\'x_n':[derivada_numerica(chute_inicial)], 
                  '|x_n - x_n-1|':[abs(solucao - chute_inicial)]}
        tabela2 = pd.DataFrame(dados2)
        tabela = tabela.append(tabela2, ignore_index = True)
        if abs(solucao - chute_inicial) < precisao:
            print("O valor de x que aproxima ao zero da função é {0:.7f}".format(solucao))
            return tabela
        i = i + 1
        chute_inicial = solucao

In [47]:
print(newton.__doc__)


    - Função: newton(f, chute_inicial, precisao, max_iteracoes)
    - Autor: Eduardo Freire Mangabeira
    - Obtém um zero de uma função pelo método de Newton.
    - Feito a partir das aulas do professor Hamilton e do livro Burden, R. L., & Faires, D. J. (1985). Numerical analysis.
    - É precisso passar uma expressão lambda para o parâmetro 'f', fiz isso para deixar o código mais geral e menos verboso.
    - Para expresões lambda, visitar: http://www.dsc.ufcg.edu.br/~pet/jornal/maio2013/materias/tutoriais.html.
    


Agora basta fornecer os valores dados pela questão para encontrar a solução. 

### Escolhendo o chute inicial

Sabemos que o zero positivo está em $[0,1]$.

Para escolher o $x_0$(chute_inicial), vamos escolher o extremo do intervalo que ao ser aplicado na função tem o mesmo sinal da segunda derivada.

$f(0) = -10 < 0$  ,   $f(1) = 4 > 0$

<br>

$f'(x) = 4x^3 + 12(x^2) + 12x + 3$
$f'(0) = 3 > 0$  ,   $f'(1) = 31 > 0$

<br>

$f''(x) = 12x^2 + 24x +12 $, temos $f''(x) > 0$ para x em $[0,1]$

<br>

Logo, a extremidade que possui o mesmo sinal da segunda derivada é $x = 1$ e será nosso chute inicial.

In [85]:
f = lambda x:  x**4 + 4*(x**3) + 6*(x**2) + 3*x - 10
chute_inicial = 1
precisao = 1/(10**4)
max_iteracoes = 10

tabela = newton(f, chute_inicial, precisao, max_iteracoes)
tabela.style.hide_index()

O valor de x que aproxima ao zero da função é 0.8555845


### Resposta

Precisamos de apenas 4 iterações para encontrar que o zero positivo da função é dado por $x = 0.8555845$

# Exercício 2

Encontre $\sqrt[4]70$ com precisão $10^{-4}$, usando o Método de Newton e 6 casas decimais nos cálculos e arredondamento.

Basta usarmos a função que escrevemos para o método de newton com os novos parâmetros!

- Obs: Para alterar o número de casas decimais exibidas no resultado final, é preciso alterar a seguinte linha:

- ```print("O valor de x que aproxima ao zero da função é {0:.7f}".format(solucao))```

- O número que está entre chaves é a quantidade de casas decimais.

### Mas onde pode estar o zero da função? 

É preciso enxergar uma função para aproximar o valor especificado. No caso:

$$f(x) = x^4 - 70$$

Sabe-se que a raiz quarta de 81 é 3, e a raiz quarta de 16 é 2.

$\sqrt[4]81 = 3$

$\sqrt[4]16 = 2$

Logo, percebe-se que a raiz quarta de 70 está no intervalo $[2,3]$.


### Escolhendo o chute inicial

Para escolher o $x_0$(chute_inicial), vamos escolher o extremo do intervalo que ao ser aplicado na função tem o mesmo sinal da segunda derivada.

$f(2) = 16 - 70 < 0$  ,   $f(3) = 81 - 70 > 0$

<br>

$f'(x) = 4x^3 $
$f'(2) = 32 > 0$  ,   $f'(3) = 108 > 0$

<br>

$f''(x) = 12x^2$, temos $f''(x) > 0$ para x em $[2,3]$

<br>

Logo, a extremidade que possui o mesmo sinal da segunda derivada é $x = 3$ e será nosso chute inicial.

In [86]:
f = lambda x: x**4 - 70
chute_inicial = 3
precisao = 1/(10**4)
max_iteracoes = 10

tabela = newton(f, chute_inicial, precisao, max_iteracoes)
tabela.style.hide_index()

O valor de x que aproxima ao zero da função é 2.8925076


### Resultado

Precisamos de apenas 3 iterações para obter que a raiz quarta de 70 é aproximadamente 2.892508.

# Exercício 3